### PR TITLE
IOTEDGE-1079 create example client that uses OAuth2 device flow

### PR DESCRIPTION
**Pull request by LimaEchoCharlie**
Output from new example:
```
./run.sh example thing/user-token -realm /example
Creating Thing dynamic-thing... Done
Requesting user code...RequestUserCode response:  {"user_code":"HqZR5F4x", ....}
Done
Requesting user token... To authorise the Thing, go to

	http://am.localtest.me:8080/am/oauth2/device/user?user_code=HqZR5F4x

Done
Access token: eyJ0eXAiOiJKV1Q....RCDF8PNioAXZx5Q
Expires in: 3599
Introspecting access token to get more information...Done
User demo has authorised the following scope(s): [publish]
 ______     __  __
/\  __ \   /\ \/ /
\ \ \/\ \  \ \  _"-.
 \ \_____\  \ \_\ \_\
  \/_____/   \/_/\/_/
```

### CODE REVIEW

File: examples/thing/user-token/main.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	fmt.Printf("Done\n")
+
+	fmt.Printf("Requesting user code...")
+	userCode, err := device.RequestUserCode("publish")
+	if err != nil {
+		return err
+	}
+	fmt.Println("Done")
+
+	fmt.Printf("Requesting user token... To authorise the Thing, go to \n\n\t%s?user_code=%s\n\n",
```
Comment (LimaEchoCharlie) [inline examples/thing/user-token/main.go:135]: I suppose I could simplify this by using the VerificationURIComplete attribute?

Comment (JacoJooste) [reply examples/thing/user-token/main.go:135]: Yep.

File: examples/thing/user-token/main.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		certBytes, err = ioutil.ReadFile(*certFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	block, _ := pem.Decode(certBytes)
+	return x509.ParseCertificates(block.Bytes)
+}
+
+// userTokenThing initialises a Thing with AM and retrieves an access token using OAuth 2.0 device flow.
```
Comment (JacoJooste) [inline examples/thing/user-token/main.go:85]:
```suggestion
// userTokenThing initialises a Thing with AM and retrieves an access token using OAuth 2.0 device authorization grant.
```

File: examples/thing/user-token/main.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+// The Thing will register and authenticate with AM and then request a user code.
+// Once the Thing is in procession of a user code, it will direct the user to authorise the token.
+// If successful, the Thing will receive a user token that contains an access token.
```
Comment (JacoJooste) [inline examples/thing/user-token/main.go:88]:
```suggestion
// If successful, the Thing will receive an access token with the user that authorised the request as the subject.
```

File: examples/thing/user-token/main.go (hunk 1 above)
Comment (JacoJooste) [inline examples/thing/user-token/main.go:135]:
```suggestion
	fmt.Printf("Requesting user access token... To authorise the request, go to \n\n\t%s?user_code=%s\n\n",
```

**Review by JacoJooste — APPROVED** [review]
Nice example 👍🏻

File: examples/thing/user-token/main.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	"os"
+
+	"github.com/ForgeRock/iot-edge/v7/pkg/builder"
+	"github.com/ForgeRock/iot-edge/v7/pkg/thing"
+)
+
+var (
+	urlString = flag.String("url", "http://am.localtest.me:8080/am", "URL of AM or Gateway")
+	realm     = flag.String("realm", "", "AM Realm")
+	audience  = flag.String("audience", "/example", "JWT audience")
```
[the remark this answers is not in the document]
Comment (LimaEchoCharlie) [reply examples/thing/user-token/main.go:37]: Just to be clear, change both of the defaults to the root realm `/`?

[the remark this answers is not in the document]
Comment (JacoJooste) [reply examples/thing/user-token/main.go:37]: Yes. I got my slashes mixed up :)

[the remark this answers is not in the document]
Comment (LimaEchoCharlie) [reply examples/thing/user-token/main.go:37]: 😄 
I'll change the other two examples as well for consistency.

[the remark this answers is not in the document]
Comment (JacoJooste) [reply examples/thing/user-token/main.go:37]: That would be great, thanks.

**Pull request merged**